GUI Fails to Build on macOS Monterey ARM

<!-- This issue tracker is only for technical issues related to Garlicoin Core.

This issue tracker is only for technical issues related to garlicoin-core.

General garlicoin questions and/or support requests and are best directed to the [garlicoin subreddit](https://www.reddit.com/r/garlicoin/) or [garlicoin discord server](https://discord.gg/garlicoin). -->

### Describe the issue

Core fails to build with QT GUI is enabled.

```
5 errors generated.
make[2]: *** [qt/garlicoin_qt-macdockiconhandler.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

### What behavior did you expect?

A successful build

### What was the actual behavior (provide screenshots if the issue is GUI-related)?

Build failed citing the error above.  With --with-gui=no it builds successfully.

### How reliably can you reproduce the issue, what are the steps to do so? -->
```
    #Install Xcode CLI tools

    xcode-select --install

    #install homebrew

    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

    #Install Macports

    https://github.com/macports/macports-base/releases/download/v2.7.2/MacPorts-2.7.2-12-Monterey.pkg

    #install dependencies

    sudo port install boost

    brew install openssl

    brew install zmq

    brew install libevent

    brew install berkeley-db@4

    brew install pkg-config

    #clone the repo

    git clone [git@github.com](mailto:git@github.com):GarlicoinOrg/Garlicoin.git

    cd Garlicoin

    #compile

    sh [autogen.sh](https://autogen.sh/)

    CPPFLAGS="-I/opt/homebrew/Cellar/libevent/2.1.12/include -I/opt/local/libexec/boost/1.76/include -I/opt/homebrew/opt/curl/include -I/opt/homebrew/opt/berkeley-db@4/include -I/opt/local/libexec/qt5/include" LDFLAGS="-L/opt/homebrew/opt/openssl@3/lib -L/opt/homebrew/Cellar/libsodium/1.0.18_1 -L/opt/homebrew/Cellar/zeromq/4.3.4 -L/opt/homebrew/Cellar/libevent/2.1.12 -L/opt/homebrew/opt/berkeley-db@4/lib -L/opt/local/libexec/qt5/lib" ./configure --with-boost=/opt/local/libexec/boost/1.76/ --disable-tests

    make -j$(nproc)
```

### What version of Garlicoin Core are you using, where did you get it (website, self-compiled, etc)? 

18, self-compiling
`HEAD is now at 9dd3913b4 Merge pull request #78 from ryan-shaw/rolling-checkpointing`

### What type of machine are you observing the error on (OS/CPU and disk type)?

```
  Model Name:	  MacBook Air
  Model Identifier:	MacBookAir10,1
  Chip:	Apple M1
  Total Number of Cores:	8 (4 performance and 4 efficiency)
  Memory:	16 GB
  System Firmware Version:	7459.141.1
  OS Loader Version:	7459.141.1

  System Version:	macOS 12.5.1 (21G83)
  Kernel Version:	Darwin 21.6.0
  Boot Volume:	Macintosh HD
  Boot Mode:	Normal
  Secure Virtual Memory:	Enabled
  System Integrity Protection:	Enabled
  Time since boot:	4 days 7:28
```

### Any extra information that might be useful in the debugging process.

https://pastebin.com/QVMiGmyQ

--- a/configure.ac
+++ b/configure.ac
@@ -444,7 +444,7 @@ case $host in
      fi
 
      AX_CHECK_LINK_FLAG([[-Wl,-headerpad_max_install_names]], [LDFLAGS="$LDFLAGS -Wl,-headerpad_max_install_names"])
-     CPPFLAGS="$CPPFLAGS -DMAC_OSX"
+     CPPFLAGS="$CPPFLAGS -DMAC_OSX -DOBJC_OLD_DISPATCH_PROTOTYPES=0"
      OBJCXXFLAGS="$CXXFLAGS"
      ;;
    *linux*)

--- a/src/qt/guiutil.cpp
+++ b/src/qt/guiutil.cpp
@@ -72,6 +72,9 @@ extern double NSAppKitVersionNumber;
 #if !defined(NSAppKitVersionNumber10_9)
 #define NSAppKitVersionNumber10_9 1265
 #endif
+#include <QProcess>
+
+void ForceActivation();
 #endif
 
 namespace GUIUtil {
@@ -405,6 +408,25 @@ bool isObscured(QWidget *w)
         && checkPoint(QPoint(w->width() - 1, w->height() - 1), w)
         && checkPoint(QPoint(w->width() / 2, w->height() / 2), w));
 }
+
+void bringToFront(QWidget* w)
+{
+    #ifdef Q_OS_MAC
+        ForceActivation();
+    #endif
+    if (w) {
+        // activateWindow() (sometimes) helps with keyboard focus on Windows
+        if (w->isMinimized()) {
+            w->showNormal();
+        } else {
+            w->show();
+        }
+        w->activateWindow();
+        w->raise();
+    }
+}
+
+
 
 void openDebugLogfile()
 {

--- a/src/qt/macdockiconhandler.mm
+++ b/src/qt/macdockiconhandler.mm
@@ -9,10 +9,9 @@
 #include <QBuffer>
 #include <QWidget>
 
-#undef slots
 #include <Cocoa/Cocoa.h>
-#include <objc/objc.h>
-#include <objc/message.h>
+#include <AppKit/AppKit.h>
+#include <objc/runtime.h>
 
 #if QT_VERSION < 0x050000
 extern void qt_mac_set_dock_menu(QMenu *);
@@ -31,18 +30,12 @@ bool dockClickHandler(id self,SEL _cmd,...) {
 }
 
 void setupDockClickHandler() {
-    Class cls = objc_getClass("NSApplication");
-    id appInst = objc_msgSend((id)cls, sel_registerName("sharedApplication"));
-    
-    if (appInst != nullptr) {
-        id delegate = objc_msgSend(appInst, sel_registerName("delegate"));
-        Class delClass = (Class)objc_msgSend(delegate,  sel_registerName("class"));
-        SEL shouldHandle = sel_registerName("applicationShouldHandleReopen:hasVisibleWindows:");
-        if (class_getInstanceMethod(delClass, shouldHandle))
-            class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
-        else
-            class_addMethod(delClass, shouldHandle, (IMP)dockClickHandler,"B@:");
-    }
+    Class delClass = (Class)[[[NSApplication sharedApplication] delegate] class];
+    SEL shouldHandle = sel_registerName("applicationShouldHandleReopen:hasVisibleWindows:");
+    if (class_getInstanceMethod(delClass, shouldHandle))
+        class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
+    else
+        class_addMethod(delClass, shouldHandle, (IMP)dockClickHandler,"B@:");
 }
 
 
@@ -131,4 +124,14 @@ void MacDockIconHandler::handleDockIconClickEvent()
     }
 
     Q_EMIT this->dockIconClicked();
+}
+
+/**
+ * Force application activation on macOS. With Qt 5.5.1 this is required when
+ * an action in the Dock menu is triggered.
+ * TODO: Define a Qt version where it's no-longer necessary.
+ */
+void ForceActivation()
+{
+    [[NSApplication sharedApplication] activateIgnoringOtherApps:YES];
 }

